### PR TITLE
Make Plan DSL independent of SymbolAllocator behavior

### PR DESCRIPTION
**Pull request by ebd2**
The plan DSL is currently dependent on the behavior of the SymbolAllocator;
creating aliases in tests requires knowing some substring of the symbol that
the SymbolAllocator generates.  This implicit dependency is brittle and is a
source for ambiguity that can lead to tests that pass incorrectly.

This change makes it so binding aliases in unit tests is done against the
rvalue that is being assigned to the symbol. This makes the DSL independent of
the SymbolAllocator's behavior.

### CODE REVIEW

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/SymbolAliases.java [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+
+    private SymbolAliases(Map<String, SymbolReference> aliases)
+    {
+        this.map = ImmutableMap.copyOf(aliases);
+    }
+
+    public SymbolAliases(SymbolAliases symbolAliases)
+    {
+        requireNonNull(symbolAliases, "symbolAliases are null");
+        this.map = new HashMap<>(symbolAliases.map);
```
Comment (martint) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/SymbolAliases.java]: `ImmutableMap.copyOf`

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/SymbolAliases.java [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+     * { "RK": SymbolReference("value3") }
+     */
+    public SymbolAliases replaceAssignments(Map<Symbol, Expression> assignments)
+    {
+        return new SymbolAliases(getUpdatedAssignments(assignments));
+    }
+
+    public static class Builder
+    {
+        Map<String, SymbolReference> tmp;
```
Comment (martint) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/SymbolAliases.java]: Call this `aliases` or `bindings` or something more descriptive than `tmp`

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/SymbolAliases.java [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+        }
+
+        /*
+         * This is supplied specifically for updateAssigments, which needs to
+         * update existing bindings that have already been added. Unless you're
+         * certain you want this behavior, you don't want it.
+         */
+        private Builder putUnchecked(Map<String, SymbolReference> aliases)
+        {
+            aliases.entrySet()
```
Comment (martint) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/SymbolAliases.java]: `tmp.putAll(aliases);`

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/Matcher.java [hunk 4]
```diff
@@ -19,5 +19,53 @@
...
+     * should be limited to tests that validate the type of the node or
+     * attributes of that type.
+     * <p>
+     * Matchers that can be applied to nodes of any typeshould return true from
+     * downMatches and do the rest of their work in upMatches.
+     *
+     * @param node The node to apply the matching tests to
+     * @return true if all matching tests pass, false otherwise
+     */
+    boolean downMatches(PlanNode node);
```
Comment (martint) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/Matcher.java]: I still think we need a better name for downMatches/upMatches

Comment (ebd2) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/Matcher.java]: Proposal here: https://github.com/prestodb/presto/pull/6584#discussion_r87866072

Comment (martint) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/Matcher.java]: That seems reasonable to me.

**Pull request closed without merging**